### PR TITLE
feat(components): [message] add vue context for message component

### PR DESCRIPTION
**Pull request by jw-foss**
- Implement context for message component to inherit
- Add test against change

Please make sure these boxes are checked before submitting your PR, thank you!

- [ ] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [ ] Make sure you are merging your commits to `dev` branch.
- [ ] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

File: packages/components/message/__tests__/message-manager.spec.tsx [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import { nextTick } from 'vue'
 import { getStyle } from '@element-plus/utils'
 import { rAF } from '@element-plus/test-utils/tick'
+import { ElMessage } from '../'
```
Comment (sxzz) [inline packages/components/message/__tests__/message-manager.spec.tsx:4]:
```suggestion
import { ElMessage } from '..'
```

File: packages/components/message/src/message.ts [hunk 2]
```diff
@@ -85,10 +85,16 @@ export interface MessageHandle {
 export type MessageParams = Partial<MessageOptions> | string | VNode
 export type MessageParamsTyped = Partial<MessageOptionsTyped> | string | VNode
 
-export type MessageFn = ((options?: MessageParams) => MessageHandle) & {
+export type MessageFn = ((
+  options?: MessageParams,
+  appContext?: null | AppContext
```
Comment (sxzz) [inline packages/components/message/src/message.ts:90]: Can we remove `null` here?
```suggestion
  appContext?: AppContext
```

File: packages/components/message/src/message-method.ts [hunk 3]
```diff
@@ -6,116 +6,121 @@ import { messageConfig } from '@element-plus/components/config-provider/src/conf
...
-    offset: verticalOffset,
-    ...options,
-    id,
-    onClose: () => {
-      close(id, userOnClose)
-    },
-  }
+    let verticalOffset = options.offset || 20
+    instances.forEach(({ vm }) => {
+      verticalOffset += (vm.el?.offsetHeight || 0) + 16
```
Comment (zouhangwithsweet) [inline packages/components/message/src/message-method.ts:56]: vm?.el

File: packages/components/message/src/message-method.ts [hunk 4]
```diff
@@ -6,116 +6,121 @@ import { messageConfig } from '@element-plus/components/config-provider/src/conf
...
+      offset: verticalOffset,
+      ...options,
+      id,
+      onClose: () => {
+        close(id, userOnClose)
+      },
+    }
+
+    let appendTo: HTMLElement | null = document.body
+    if (options.appendTo instanceof HTMLElement) {
```
Comment (zouhangwithsweet) [inline packages/components/message/src/message-method.ts:75]: should we have an util fn to check Element like `isHtmlElement()`?

File: packages/components/message/src/message.ts (hunk 2 above)
Comment (JeremyWuuuuu) [reply packages/components/message/src/message.ts:90]: No, it is simply mapping for `Vue`'s typing

File: packages/components/message/src/message-method.ts (hunk 3 above)
Comment (JeremyWuuuuu) [reply packages/components/message/src/message-method.ts:56]: Nothing changed here, this is just formatting.

File: packages/components/message/src/message-method.ts (hunk 4 above)
Comment (JeremyWuuuuu) [reply packages/components/message/src/message-method.ts:75]: Yes we do, but that was in `components/images` we should extract that out.

File: packages/components/message/src/message.ts (hunk 2 above)
Comment (sxzz) [reply packages/components/message/src/message.ts:90]: I didn't get your point.

Comment (JeremyWuuuuu) [reply packages/components/message/src/message.ts:90]: image

Comment (sxzz) [reply packages/components/message/src/message.ts:90]: OK.

**Review by sxzz — APPROVED** [review]

File: packages/components/message/src/message-method.ts [hunk 5]
```diff
@@ -6,116 +6,121 @@ import { messageConfig } from '@element-plus/components/config-provider/src/conf
...
+            ).visible = false),
+        }
       }
     }
-  }
 
-  if (typeof options === 'string' || isVNode(options)) {
-    options = { message: options }
-  }
+    if (typeof options === 'string' || isVNode(options)) {
```
Comment (btea) [inline packages/components/message/src/message-method.ts:50]:
```suggestion
    if (isString(options) || isVNode(options)) {
```

Comment (JeremyWuuuuu) [reply packages/components/message/src/message-method.ts]: Done!

**Pull request merged**